src/haxelib/api/Installer.hx

```diff
@@ -0,0 +1,852 @@
+		final library = ProjectName.ofString(info.name);
+
+		final versions = [for (v in info.versions) v.name];
+
+		if (versions.length == 0)
+			throw new InstallationException('The library $library has not yet released a version');
+
+		final version = getLatest(versions);
+		downloadAndInstall(library, version);
+
+		scope.setVersion(library, version);
+
+		userInterface.log('  Current version is now $version');
+		userInterface.log("Done");
+
+		handleDependencies(library, version);
+	}
+
+	/**
+		Installs `version` of `library` from the haxelib server.
+
+		If `forceSet` is set to true and the installer is running with
+		a global scope, the new version is always set as the current one.
+	 **/
+	public function installFromHaxelib(library:ProjectName, version:SemVer, forceSet:Bool = false) {
+		final info = Connection.getInfo(library);
+		// get correct name capitalization so that the logged output is correct
+		final library = ProjectName.ofString(info.name);
+
+		final versions = [for (v in info.versions) v.name];
+
+		if (versions.length == 0)
+			throw new InstallationException('The library $library has not yet released a version');
+		if (!versions.contains(version))
+			throw new InstallationException('No such version $version for library $library');
+
+		downloadAndInstall(library, version);
+
+		if (scope.isLocal || version == getLatest(versions) || !scope.isLibraryInstalled(library) || forceSet) {
+			scope.setVersion(library, version);
+			userInterface.log('  Current version is now $version');
+		}
+		userInterface.log("Done");
+
+		handleDependencies(library, version);
+	}
+
+	/**
+		Installs `library` from a git or hg repository (specified by `id`)
+
+		`vcsData` contains information on the source repository
+		and the requested state.
+	**/
+	public function installVcsLibrary(library:ProjectName, id:VcsID, vcsData:VcsData) {
+		installVcs(library, id, vcsData);
+
+		library = getName(library, id, vcsData.subDir);
+
+		scope.setVcsVersion(library, id, vcsData);
+
+		if (vcsData.subDir != null) {
+			final path = scope.getPath(library);
+			userInterface.log('  Development directory set to $path');
+		} else {
+			userInterface.log('  Current version is now $id');
+		}
+		userInterface.log("Done");
+
+		handleDependenciesVcs(library, id, vcsData.subDir);
+	}
+
+	/**
+		Updates `library` to the newest version on the haxelib server,
+		or pull latest changes with git or hg.
+	**/
+	public function update(library:ProjectName) {
+		try {
+			updateIfNeeded(library);
+		} catch (e:AlreadyUpToDate) {
+			userInterface.log(e.toString());
+		} catch (e:VcsUpdateCancelled) {
+			return;
+		}
+	}
+
+	/**
+		Updates all libraries in the scope.
+
+		If a library update fails, it is skipped.
+	**/
+	public function updateAll():Void {
+		final libraries = scope.getLibraryNames();
+		var updated = false;
+
+		for (library in libraries) {
+			userInterface.log('Checking $library');
+			try {
```

Comment:
Surely this should call `update` instead of duplicating.

Reply:
The shared behaviour is in `updateIfNeeded()`. `update()` and `updateAll` have slightly differently handling if the library it tried to update is already up to date.

Reply:
The only difference I'm seeing is that `update` logs...

Reply:
Yes, they log different things if a library is already up to date. And updateAll continues even if one library update throws an exception. But this logging stuff can't be merged together as it is different for the two, so the behaviour that can be merged has already been put into the updateIfNeeded method.